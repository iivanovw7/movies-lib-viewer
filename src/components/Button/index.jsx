/**
 * Module contains button component
 * @module components/Button
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import * as R from 'ramda';
import React from 'react';

import { isUndefined } from '../../utils/helpers';

import StyledButton from './StyledButton';
import StyledLink from './StyledLink';

/**
 * Creates styled button component.
 * @name Components/Button
 * @method
 *
 * @param {Object.<module:components/Button~propTypes>} props
 *  contains component props
 *  @see {@link module:components/Button~propTypes}
 * @return {Node} React component with children.
 * @constructor
 */
const Button = (props) => {
  const { variant, children, styling, onClick, index } = props;

  /* eslint-disable react/jsx-props-no-spreading */
  const ButtonElement = R.ifElse(
    R.pipe(R.prop('onClick'), isUndefined),
    R.always(<StyledLink {...props} />),
    R.always(<StyledButton {...props} />),
  );
  /* eslint-enable react/jsx-props-no-spreading */

  return (
    <ButtonElement data-index={index} variant={variant} styling={styling} onClick={onClick}>
      {children}
    </ButtonElement>
  );
};

/**
 * @name propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {string} [props.variant = 'primary'] - color variant string.
 * @property {string | Image} props.children - button children element or text.
 * @property {Array.<string>} [props.styling = []] - image additional styles.
 * @property {Function} props.onClick - click handler.
 * @property {string} props.href - target link.
 * @property {string} [props.target = '_blank'] - target prop.
 * @property {boolean} [props.hidden = false] - if element should be hidden.
 * @property {number} [props.index = 0] - index which is passed in `data-index` attribute.
 * @return {Array} React propTypes
 */
Button.propTypes = {
  variant: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.node])),
  ]),
  styling: PropTypes.array,
  onClick: PropTypes.func,
  href: PropTypes.string,
  target: PropTypes.string,
  hidden: PropTypes.bool,
  index: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Button.defaultProps = {
  target: '_blank',
  hidden: false,
  index: 0,
};

export default Button;
