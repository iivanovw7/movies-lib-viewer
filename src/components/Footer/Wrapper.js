/**
 * Module contains application footer wrapper
 * @module components/Footer/Wrapper
 * @author Igor Ivanov
 */
import styled from 'styled-components';

import { base } from '../../config/styles';
import { zIndex } from '../../style/mixins';

const Wrapper = styled.footer`
  bottom: 0;
  display: flex;
  flex-direction: row;
  height: ${base.footerHeight}em;
  justify-content: center;
  position: fixed;
  width: 100%;
  z-index: ${zIndex.Footer};
`;

export default Wrapper;
