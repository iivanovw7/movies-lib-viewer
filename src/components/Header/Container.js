/**
 * Module contains application header container
 * @module components/Header/Container
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const Container = styled.div`
  align-items: center;
  display: flex;
  flex-direction: row;
  height: 100%;
  justify-content: flex-start;
  width: 100%;
`;

export default Container;
