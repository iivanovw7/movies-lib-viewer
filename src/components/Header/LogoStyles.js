/**
 * Module contains additional styles used in logo image.
 * @module components/TopBar/LogoLogoStyles
 * @author Igor Ivanov
 */
import { css } from 'styled-components';

const LogoStyles = css`
  margin-left: 0.8em;

  svg {
    height: 3rem;
    max-height: 48px;
    max-width: 48px;
    width: 3rem;
  }
`;

export default LogoStyles;
