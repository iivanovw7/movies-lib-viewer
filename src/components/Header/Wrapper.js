/**
 * Module contains application header wrapper
 * @module components/Header/Wrapper
 * @author Igor Ivanov
 */
import styled from 'styled-components';

import { base } from '../../config/styles';
import { zIndex } from '../../style/mixins';
import { headerBg } from '../../style/theme/background';

const Wrapper = styled.header`
  background-color: ${headerBg};
  display: column;
  flex-direction: row;
  height: ${base.headerHeight}em;
  justify-content: center;
  margin-top: ${base.topBarHeight}em;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: ${zIndex.Header};

  &::after {
    background: url(${(props) => props.bg});
    background-repeat: no-repeat;
    background-size: cover;
    bottom: 0;
    content: '';
    left: 0;
    opacity: 0.3;
    position: absolute;
    right: 0;
    top: 0;
    z-index: -1;
  }
`;

export default Wrapper;
