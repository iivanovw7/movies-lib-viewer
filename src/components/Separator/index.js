/**
 * Module contains Separator component
 * @module components/Separator
 * @author Igor Ivanov
 */
import { linearGradient } from 'polished';
import styled from 'styled-components';

import { colorSet } from '../../config/styles';
import { zIndex } from '../../style/mixins';

const { gradients } = colorSet;

const Separator = styled.div`
  height: ${(props) => props.height || 0.1}em;
  ${linearGradient({
    colorStops: [`${gradients.first} 0%`, `${gradients.second} 50%`, `${gradients.third} 95%`],
    toDirection: 'to right',
    fallback: gradients.first,
  })};
  width: 100%;
  z-index: ${zIndex.Separator};
`;

export default Separator;
