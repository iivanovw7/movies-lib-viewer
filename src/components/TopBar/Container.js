/**
 * Module contains application TopBar container
 * @module components/TopBar/Container
 * @author Igor Ivanov
 */
import styled from 'styled-components';

import { base } from '../../config/styles';
import { zIndex } from '../../style/mixins';
import { topBarBg } from '../../style/theme/background';
import { topBarBorderColor } from '../../style/theme/borders';

const Container = styled.div`
  align-items: center;
  background-color: ${topBarBg};
  border-bottom: 1px solid ${topBarBorderColor};
  display: flex;
  flex-direction: row;
  height: ${base.topBarHeight}em;
  justify-content: space-between;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: ${zIndex.TopBar};
`;

export default Container;
