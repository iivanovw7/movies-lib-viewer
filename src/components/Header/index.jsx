/**
 * Module contains application header
 * @module components/Header
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import React from 'react';

import HeaderBg from '../../../assets/img/loaderBg.jpg';
import { ReactComponent as APPLogo } from '../../../assets/svg/film.svg';
import Button from '../Button';
import Separator from '../Separator';

import Container from './Container';
import LogoStyles from './LogoStyles';
import Wrapper from './Wrapper';

/**
 * Creates application header component.
 * @name Components/Header
 * @method
 * @param {Object.<module:components/Header~propTypes>} props
 *  contains component props
 *  @see {@link module:components/Header~propTypes}
 * @constructor
 * @return {Node} React component with children.
 */
function Header(props) {
  const { children } = props;

  return (
    <Wrapper bg={HeaderBg}>
      <Container>
        <Button variant="primary" href="/" target="_self" styling={LogoStyles}>
          <APPLogo fill="currentColor" />
        </Button>
        {children}
      </Container>
      <Separator />
    </Wrapper>
  );
}

/**
 * @name propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {string | Image} props.children
 *    header children elements.
 * @return {Array} React propTypes
 */
Header.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.node])).isRequired,
  ]),
};

export default Header;
