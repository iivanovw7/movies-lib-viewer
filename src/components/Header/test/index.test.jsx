/**
 * Module contains Header component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';

import HeaderBg from '../../../../assets/img/loaderBg.jpg';
import ChildMock from '../../../__mocks__/childMock';
import { testName, mountWithTheme } from '../../../__utils__/utils';
import { colorSet } from '../../../config/styles';
import Button from '../../Button';
import Separator from '../../Separator';
import Container from '../Container';
import Header from '../index';
import LogoStyles from '../LogoStyles';
import Wrapper from '../Wrapper';

/* eslint-disable require-jsdoc, react/jsx-props-no-spreading */
// prettier-ignore
describe(testName('Header', 'component test suite'), function HeaderTestSuite() {
  const Composition = (props, children, shouldMount = false) => {
    const node = (<Header {...props}>{children}</Header>);

    return shouldMount ? mount(node) : shallow(node);
  };

  it('Should match snapshot', () => {
    const component = Composition({}, ChildMock.component);

    expect(component)
      .toMatchSnapshot();
  });

  it('Should render Wrapper', () => {
    const component = Composition({}, ChildMock.component);

    expect(component.find(Wrapper))
      .toHaveLength(1);
  });

  it('Should render Container', () => {
    const component = Composition({}, ChildMock.component);

    expect(component.find(Container))
      .toHaveLength(1);
  });

  it('Should render Separator', () => {
    const component = Composition({}, ChildMock.component);

    expect(component.find(Separator))
      .toHaveLength(1);
  });

  it('Should render child mock', () => {
    const component = Composition({}, ChildMock.component);
    const child = component.find(`#${ChildMock.id}`);

    expect(child)
      .toHaveLength(1);

    expect(child.text())
      .toEqual(ChildMock.child);
  });

  it('Should verify bg prop passing on shallow', () => {
    const component = Composition({}, ChildMock.component);

    expect(component.prop('bg'))
      .toBe(HeaderBg);

  });

  /*
    Had to downgrade versions to styled-components@4.4.1 and jest-styled-components@6.0.0
      as last versions are no compatible with styled theming, details:
      https://github.com/styled-components/jest-styled-components/issues/234
    TODO upgrade versions after issue is resolved
   */
  it('Should verify bg style rule on mount', () => {
    const component = mountWithTheme(<Header>{ChildMock.component}</Header>);

    expect(component.find(Wrapper))
      .toHaveStyleRule('display', 'column');

    expect(component)
      .toHaveStyleRule('background-color', `${colorSet.dark.headerBackgroundPrimary}`);

  });

  it(testName('Button', 'Should render <Button /> with props'), () => {
    const component = Composition({}, ChildMock.component);
    const button = component.find(Button);

    expect(button)
      .toHaveLength(1);

    expect(button.prop('target'))
      .toEqual('_self');

    expect(button.prop('href'))
      .toEqual('/');

    expect(button.prop('styling'))
      .toBe(LogoStyles);
  });

});
/* eslint-enable require-jsdoc, react/jsx-props-no-spreading */
