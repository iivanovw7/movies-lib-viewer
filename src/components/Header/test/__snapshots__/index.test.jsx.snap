// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mHeader[39m]: component test suite Should match snapshot 1`] = `
<Wrapper
  bg="test-file-stub"
>
  <Container>
    <Button
      hidden={false}
      href="https://www.themoviedb.org/?language=en-US"
      index={0}
      styling={
        Array [
          "margin-left:0.8em;svg{height:3rem;max-height:48px;max-width:48px;width:3rem;}",
        ]
      }
      target="_blank"
      variant="primary"
    >
      <ForwardRef
        fill="currentColor"
      />
    </Button>
    <div
      id="child-mock-id"
    >
      child-mock-id
    </div>
  </Container>
  <Separator />
</Wrapper>
`;
